receiver/zookeeperreceiver/scraper.go

```diff
@@ -15,32 +15,192 @@
 package zookeeperreceiver
 
 import (
+	"bufio"
 	"context"
+	"errors"
+	"fmt"
+	"net"
+	"regexp"
+	"strconv"
+	"time"
 
 	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/collector/consumer/simple"
 	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/zookeeperreceiver/internal/metadata"
+)
+
+var zookeeperFormatRE = regexp.MustCompile(`(^zk_\w+)\s+([\w\.\-]+)`)
+
+const (
+	mntrCommand = "mntr"
 )
 
 type zookeeperMetricsScraper struct {
-	cfg *Config
-}
+	logger *zap.Logger
+	config *Config
+	cancel context.CancelFunc
 
-func newZookeeperMetricsScraper(_ *zap.Logger, cfg *Config) (*zookeeperMetricsScraper, error) {
-	return &zookeeperMetricsScraper{cfg: cfg}, nil
+	// For mocking.
+	closeConnection       func(net.Conn) error
+	setConnectionDeadline func(net.Conn, time.Time) error
+	sendCmd               func(net.Conn, string) (*bufio.Scanner, error)
 }
 
 func (z *zookeeperMetricsScraper) Name() string {
-	return z.cfg.Name()
+	return typeStr
+}
+
+func newZookeeperMetricsScraper(logger *zap.Logger, config *Config) (*zookeeperMetricsScraper, error) {
+	_, _, err := net.SplitHostPort(config.TCPAddr.Endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	if config.Timeout <= 0 {
+		return nil, errors.New("timeout must be a positive duration")
+	}
+
+	return &zookeeperMetricsScraper{
+		logger:                logger,
+		config:                config,
+		closeConnection:       closeConnection,
+		setConnectionDeadline: setConnectionDeadline,
+		sendCmd:               sendCmd,
+	}, nil
 }
 
 func (z *zookeeperMetricsScraper) Initialize(_ context.Context) error {
 	return nil
 }
 
 func (z *zookeeperMetricsScraper) Close(_ context.Context) error {
+	z.cancel()
 	return nil
 }
 
-func (z *zookeeperMetricsScraper) Scrape(_ context.Context, _ string) (pdata.ResourceMetricsSlice, error) {
-	return pdata.ResourceMetricsSlice{}, nil
+func (z *zookeeperMetricsScraper) Scrape(ctx context.Context, _ string) (pdata.ResourceMetricsSlice, error) {
+	var ctxWithTimeout context.Context
+	ctxWithTimeout, z.cancel = context.WithTimeout(ctx, z.config.Timeout)
+
+	conn, err := z.config.Dial()
+	if err != nil {
```

Comment:
Any plans to track all members of a cluster and try a different one if connection fails?

Reply:
Currently the idea is to be able to collect metrics only from the host the collector is running on. In cases where there are multiple members within a cluster, each instance is expected to have a `zookeeper` receiver running to collect metrics. If a connection fails for some reason, it'll be tried again in the subsequent scrape interval.